### PR TITLE
[CompilerCaching] Path Remapping for canonicalization

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
Teach dependency scanner to canonicalize the path for better cache hit and also teach swift-frontend to restore the path to the actual on disk path after canonicalization.

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

File: include/swift/Option/Options.td [hunk 1]
```diff
@@ -1853,6 +1853,22 @@ def clang_scanner_module_cache_path : Separate<["-"], "clang-scanner-module-cach
   Flags<[FrontendOption, DoesNotAffectIncrementalBuild, ArgumentIsPath]>,
   HelpText<"Specifies the Clang dependency scanner module cache path">;
 
+def scanner_prefix_map : Separate<["-"], "scanner-prefix-map">,
+  Flags<[FrontendOption, NewDriverOnlyOption]>,
+  HelpText<"Remap paths reported by dependency scanner">, MetaVarName<"<prefix=replacement>">;
+
+def scanner_prefix_map_sdk : Separate<["-"], "scanner-prefix-map-sdk">,
```
[the remark this answers is not in the document]
Comment (cachemeifyoucan) [reply include/swift/Option/Options.td:1860]: This is for `swift-driver`. SwiftDriver knows about SDK path and toolchain path, and it is going to convert that information to `-scanner-prefix-map` option for swift-frontend.

**Review by artemcm — APPROVED** [review]

**Comment by cachemeifyoucan** [comment]
Rebase after merge conflict.

@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Pull request merged**